libraries/base/Data/Semigroup.hs

```diff
@@ -18,21 +18,37 @@
 -- Maintainer  :  libraries@haskell.org
 -- Stability   :  provisional
 -- Portability :  portable
+-- 
+-- A type @a@ is a 'Semigroup' if it provides an associative function ('<>')
+-- that lets you combine any two values of type @a@ into one. 
 --
--- In mathematics, a semigroup is an algebraic structure consisting of a
--- set together with an associative binary operation. A semigroup
--- generalizes a monoid in that there might not exist an identity
--- element. It also (originally) generalized a group (a monoid with all
--- inverses) to a type where every element did not have to have an inverse,
--- thus the name semigroup.
+-- ==== __Examples__
+-- >>> sconcat (1 :| [2, 3, 4]) :: Min Int
+-- Min {getMin = 1}
+-- >>> sconcat (1 :| [2, 3, 4]) :: Max Int
+-- Max {getMax = 4}
+-- >>> "foo" <> "bar"
+-- "foobar"
+-- >>> Last 1 <> Last 2 <> Last 3
+-- Last {getLast 3}
+-- >>> First 1 <> First 2 <> First 3
+-- First {getFirst = 1}
```

Comment:
Examples <3

Reply:
Maybe it would be nice to add a word why `NonEmpty` appears in the class instance (if it's not too hard to understand)?
```
@NonEmpty a@ is a semigroup for each @a@ and this the free semigroup, i.e. there exists 
@
  foldMapSemigroup :: Semigroup s => (a -> s) -> NonEmpty a -> s
@
@sconcat@ is then just @foldMapSemigroup id@, moreover @foldMapSemigroup@ is analogous to 
@foldMap :: Monoid m => (a -> m) -> [a] -> m@.
```

Unfortunately there is no `foldMapSemigroup` in `Semigroup`, maybe it's worth adding it?
```
foldMapSemigroup :: Semigroup s => (a -> s) => NonEmpty a -> s
foldMapSemigroup f = sconcat . fmap f
```

Reply:
Anyway `sconcat` will be there and it might be good to know where it comes from, plus you learn a pattern that is repeating else where (and it is repeating itself all over Haskell ecosystem).